diff --git a/crates/nu-engine/src/env.rs b/crates/nu-engine/src/env.rs
index 07c7a7346..4e8d44ed1 100644
--- a/crates/nu-engine/src/env.rs
+++ b/crates/nu-engine/src/env.rs
@@ -271,17 +271,26 @@ pub fn find_in_dirs_env(
             return None;
         }
 
-        lib_dirs?
-            .as_list()
-            .ok()?
-            .iter()
-            .map(|lib_dir| -> Option<PathBuf> {
-                let dir = lib_dir.to_path().ok()?;
-                let dir_abs = canonicalize_with(dir, &cwd).ok()?;
-                canonicalize_with(filename, dir_abs).ok()
-            })
-            .find(Option::is_some)
-            .flatten()
+        let lib_dirs = lib_dirs?;
+        if let Ok(dirs) = lib_dirs.as_list() {
+            dirs.iter()
+                .map(|lib_dir| -> Option<PathBuf> {
+                    let dir = lib_dir.to_path().ok()?;
+                    let dir_abs = canonicalize_with(dir, &cwd).ok()?;
+                    canonicalize_with(filename, dir_abs).ok()
+                })
+                .find(Option::is_some)
+                .flatten()
+        } else if let Ok(dirs) = lib_dirs.as_str() {
+            std::env::split_paths(dirs)
+                .filter_map(|lib_dir| {
+                    let dir_abs = canonicalize_with(&lib_dir, &cwd).ok()?;
+                    canonicalize_with(filename, dir_abs).ok()
+                })
+                .next()
+        } else {
+            None
+        }
     };
 
     let lib_dirs = dirs_var.and_then(|var_id| engine_state.get_var(var_id).const_val.as_ref());
diff --git a/crates/nu-parser/src/parse_keywords.rs b/crates/nu-parser/src/parse_keywords.rs
index 330ff2e9b..d2e71fda4 100644
--- a/crates/nu-parser/src/parse_keywords.rs
+++ b/crates/nu-parser/src/parse_keywords.rs
@@ -4093,22 +4093,26 @@ pub fn find_in_dirs(
                 if let Some(lib_dirs) =
                     dirs_env.and_then(|dirs_env| working_set.get_env_var(dirs_env))
                 {
-                    if let Ok(dirs) = lib_dirs.as_list() {
-                        for lib_dir in dirs {
-                            if let Ok(dir) = lib_dir.to_path() {
-                                // make sure the dir is absolute path
-                                if let Ok(dir_abs) = canonicalize_with(dir, actual_cwd)
-                                    && let Ok(path) = canonicalize_with(filename, dir_abs)
-                                {
-                                    return Some(path);
-                                }
-                            }
-                        }
-
-                        None
+                    let dir_candidates: Vec<PathBuf> = if let Ok(dirs) = lib_dirs.as_list() {
+                        dirs.iter()
+                            .filter_map(|lib_dir| lib_dir.to_path().ok())
+                            .collect()
+                    } else if let Ok(dirs) = lib_dirs.as_str() {
+                        std::env::split_paths(dirs).collect()
                     } else {
-                        None
+                        return None;
+                    };
+
+                    for lib_dir in dir_candidates {
+                        // Make sure the dir is absolute path.
+                        if let Ok(dir_abs) = canonicalize_with(lib_dir, actual_cwd)
+                            && let Ok(path) = canonicalize_with(filename, dir_abs)
+                        {
+                            return Some(path);
+                        }
                     }
+
+                    None
                 } else {
                     None
                 }
